### DISCUSSION TITLE
Is it possible to disable a rule only inside a specific LaTeX environment?

### DISCUSSION THREAD

**Discussion by zertrin** [valentjn/vscode-ltex#329] [Q&A] [answered]
I'm mainly thinking about disabling the `UPPERCASE_SENTENCE_START` rule, but only inside `itemize` environments, as I get many false positive for this one, but I guess this may be useful for other usages. I still want the rule active outside the environment.

Currently, the way to add this to the allowlist this is linked with the actual sentence. But there may be many instances and the text might evolve over time, rendering the allowlist rules obsolete.

Ideally, I would love to have a way to say "disable rule XYZ inside LaTeX environment ABC".

Is it already possible somehow? (yes maybe through a crafted regex, but I'd prefer avoiding using regexes to match environment blocks, if possible)

Otherwise, would this be a valid feature request?

**Comment by valentjn** [comment, marked as answer]
This sounds like a pretty specific feature to me. Do you have more examples of rules for which it might be useful to disable them in a specific environment? And can you give some examples where you get `UPPERCASE_SENTENCE_START` in an `itemize` environment (is it right after `\item`, etc.)? Maybe we can add some logic that works just for that combination. I'm a bit reluctant to implement general features with only very few use cases, as every additional feature increases UX complexity.

**Reply by zertrin** [reply]
It is indeed mainly in the following environments: `enumerate`, `itemize`, `description` as well as in tables (`tabular`, `tabu`).

My thesis contains many cases of lists/tables where each element is not a real sentence, but instead a list of some short technical words.

I was suggesting a generic approach because I wasn't sure whether this was an easy/harder task to implement. If a specific case is easier to implement on your side, maybe that way is better.

I guess covering at least the `\item` use case would solve 70% of the troubles I have with this rule.

**Reply by valentjn** [reply]
Thanks, but you didn't answer this question: &ldquo;And can you give some examples where you get `UPPERCASE_SENTENCE_START` in an itemize environment (is it right after `\item`, etc.)?&rdquo; For example, I don't get any diagnostics here:

```latex
This is a list:

\begin{itemize}
  \item apples
  \item oranges
  \item pears
\end{itemize}
```

If you want to request a feature, even if you're unsure if it's already possible or if it's a reasonable feature request, please open a feature request issue by filling out the corresponding template (as the template will specifically ask you to provide real-world examples). We can continue the discussion there.

**Reply by zertrin** [reply]
Sorry for delay in replying, been too busy recently...

Something must have already changed between the time I opened this discussion (back in beginning of June) and now, as it seems that I do not really have the issue as often as before.

I was even able to remove the exceptions I previously added via quick fix to `.vscode/ltex.hiddenFalsePositives.en-GB.txt` and the warning didn't reappear even if the text is still there.

Maybe unrelated, but I recently noticed that there is a difference between:

```latex
This is the result from the balance between:
\begin{itemize}
    \item the redacted content, and
    \item the redacted content.
\end{itemize}
```
and
```latex
This is the result from the balance between:

\begin{itemize}
    \item the redacted content, and
    \item the redacted content.
\end{itemize}
```

Only in the second case does the `UPPERCASE_SENTENCE_START` warning trigger.

That is fine, as the first case is anyway more logical.

Moreover, in situations like:
```latex
The scope of redacted content includes:

\begin{itemize}
    \item to simulate the redacted content
    \item to enable the redacted content
    \item from the simulation results, to derive the redacted content
\end{itemize}
```
there is no warning.

So, in the end, it's unclear to me under what exact conditions the `UPPERCASE_SENTENCE_START` warning gets triggered in relation to `itemize` lists.

**Reply by valentjn** [reply]
LT<sub>E</sub>X basically converts L<sup>A</sup>T<sub>E</sub>X code to plaintext (removing the markup) and then feeds that plaintext to LanguageTool (LT). You can see the first 100 characters of the plaintext in the `LTeX Language Server` log under *View* &#x203a; *Output*. Whether there is a diagnostic or not is decided by LT. LT<sub>E</sub>X does not alter the results obtained by LT.

If you upgraded to LT<sub>E</sub>X 10.5.0 before your last comment (it was released 8 days before), then you're using a different LT version than before (5.4 vs. 5.3). LT regularly changes its matching rules, which might well be the reason for the different behavior. Unfortunately, the [changelog of LT](https://github.com/languagetool-org/languagetool/blob/v5.4/languagetool-standalone/CHANGES.md#54-2021-06-25) is poorly maintained, so it's hard to verify this.

If the `itemize` list belongs to the same sentence as the previous line (or the next one), you should not add an empty line between them, because this designates a paragraph break in L<sup>A</sup>T<sub>E</sub>X. This would be typographically bad, as a single sentence must not span over multiple paragraphs. Therefore, LT<sub>E</sub>X and LT actually correctly deal with the first two examples. In addition, inserting an empty line might change how the document is typeset, e.g., an empty line after the list will cause the next sentence to have a paragraph indent via `\parindent`, making the paragraph break visible to the reader's eye.

In the third example, the existence of a diagnostic depends on whether LT deems the items to be full sentences. This depends, as mentioned above, on LT's internal rules, and is out of LT<sub>E</sub>X's scope. In this concrete case, I would prefer incorporating the items into one sentence like in the first example: Either
* the items are short enough (this is the case in all of your examples) so that the resulting sentence is not too convoluted, or
* the items are longer. In this case, you should start a new sentence at each new item, or break the items into multiple sentences, just like I'm doing right now.
